### ISSUE TITLE
TS error cloneElement children

### ISSUE THREAD

**Issue by Frikki** [preactjs/preact#2428]
### Reproduction

https://codesandbox.io/s/strange-wilson-w6k4p

### Steps to reproduce

- Use `preact@10.3.4`
- Pass `props.children` to `cloneElement`.

### Expected Behavior

As per the [documentation example](https://preactjs.com/guide/v10/api-reference/#cloneelement), this should be valid. Thus, no TS errors should occur.

### Actual Behavior

TS reports the following error:

```
No overload matches this call.
  Overload 1 of 2, '(vnode: VNode<any>, props?: any, ...children: ComponentChildren[]): VNode<any>', gave the following error.
    Argument of type 'ComponentChildren' is not assignable to parameter of type 'VNode<any>'.
      Type 'undefined' is not assignable to type 'VNode<any>'.
  Overload 2 of 2, '(vnode: VNode<any>, props?: any, ...children: ComponentChildren[]): VNode<any>', gave the following error.
    Argument of type 'ComponentChildren' is not assignable to parameter of type 'VNode<any>'.
      Type 'undefined' is not assignable to type 'VNode<any>'.ts(2769)
```

**Comment by 38elements** [comment]
This works.

```ts
import { h, FunctionalComponent, ComponentChildren, cloneElement, render, VNode } from 'preact'

interface Props {
  children: ComponentChildren
}

const Linkout: FunctionalComponent<Props> = function (props: Props) {
  return cloneElement(props.children as VNode, {target: '_blank'})
}


render(<Linkout><a href="/">home</a></Linkout>, document.body)
```

**Comment by Frikki** [comment]
> This works.
> 
> ```ts
> import { h, FunctionalComponent, ComponentChildren, cloneElement, render, VNode } from 'preact'
> 
> interface Props {
>   children: ComponentChildren
> }
> 
> const Linkout: FunctionalComponent<Props> = function (props: Props) {
>   return cloneElement(props.children as VNode, {target: '_blank'})
> }
> 
> 
> render(<Linkout><a href="/">home</a></Linkout>, document.body)
> ```

Yes, but casting `ComponentChildren` as `VNode` is a hack and doesn’t address the underlying issue. However, some might find the hack useful.

**Comment by ForsakenHarmony** [comment]
Would you be able to submit a PR fixing it?

**Comment by Frikki** [comment]
> Would you be able to submit a PR fixing it?

I could give it a shot.

**Comment by ddayguerrero** [comment]
Took at stab at resolving this. In fact, I believe [my changes](https://github.com/preactjs/preact/pull/2388) caused this type error. Hope you don't mind!

**Comment by kim3er** [comment]
I'm not convinced this an issue with the typing on `cloneElement`. `cloneElement` correctly only accepts a `VNode`.

If creating a wrapping component like `<Linkout />`, the type of `children` should be narrowed. Which has the same effect as @38elements's suggestion.

```
function Linkout(props: Omit<JSXInternal.HTMLAttributes, 'children'> & { children: VNode }) {
  // add target="_blank" to the link:
  return cloneElement(props.children, { target: '_blank' });
}
```

For additional type safety, I'd add generics to the `cloneElement` definition. With that in place `{ foo: '_blank' }` would throw an error.

```
export function cloneElement<RefType extends EventTarget>(
  vnode: VNode<RefType>,
  props?: JSXInternal.HTMLAttributes<RefType>,
  ...children: ComponentChildren[]
): VNode<RefType>;

function Linkout<RefType extends HTMLAnchorElement>(props: Omit<JSXInternal.HTMLAttributes<RefType>, 'children'> & { children: VNode<RefType> }) {
  // add target="_blank" to the link:
  return cloneElement<RefType>(props.children, { target: '_blank' });
}
```

If you're happy with my assessment, I'd happily create a PR for the `cloneElement` typing.

**Comment by JoviDeCroock** [comment]
Closing this as working as intended. `ComponentChildren` can contain values that are valid children but are not VNodes, including strings, numbers, arrays, `null`, and `undefined`. `cloneElement` can only clone a VNode, so accepting `ComponentChildren` would allow values the runtime cannot clone.

This matches React's distinction between `ReactNode` and `ReactElement`: `cloneElement` requires an element there as well. Components that require one cloneable child should type `children` as `VNode`, or narrow broader children with `isValidElement()` before calling `cloneElement`.
